src/mpi4py/MPI/asdlpack.pxi

```diff
@@ -200,7 +198,7 @@ cdef int Py_GetDLPackBuffer(object obj, Py_buffer *view, int flags) except -1:
 
         fixnull = (buf == NULL and size == 0)
         if fixnull: buf = &fixnull
-        PyBuffer_FillInfo(view, obj, buf, size, readonly, flags)
+        PyBuffer_FillInfo(view, capsule, buf, size, readonly, flags)
```

Comment:
This is not right! The `PyBuffer_FillInfo` should be called with the "exporter" and a reference is held in the `view` struct. The capsule is not the exporter. Moreover, if the `view` struct would hold a reference to the capsule as you are doing, then calling `manager.deleter` is not appropriate. 

Please note that these "getbuffer" routines also support the `memory` class, which a mpi4py's stripped-down version of Python's `memoryview`. For that support to keep working properly, `PyBuffer_FillInfo` should be passed `obj` and not `capsule`.

Reply:
The problem I was trying to fix is that the `deleter` gets called too early below. Here we are preparing the buffer to be consumed by MPI. The `deleter` should be called *after* the MPI operation is done, not before. I made the buffer hold a reference to `capsule` so that when the buffer (`view`) is done, the capsule's destructor (in which a sane implementation would call `deleter`) can be called. Then, we can remove the `deleter` call below.

> Please note that these "getbuffer" routines also support the `memory` class, which a mpi4py's stripped-down version of Python's `memoryview`. For that support to keep working properly, `PyBuffer_FillInfo` should be passed `obj` and not `capsule`.

I missed this part. This may have to do with the errors I am seeing. Could you elaborate?

Reply:
What's the responsibility of the deleter? Is the exporter supposed to keep track of how many alive capsules have been produced? Can you tell me what's the specific issue/problem with calling the deleter before the MPI call? I always assumed that the only purpose of DLPack was to pass buffer metadata around, not keep things "alive". Why isn't enough with keeping the exporter alive?

Reply:
> Can you tell me what's the specific issue/problem with calling the deleter before the MPI call? I always assumed that the only purpose of DLPack was to pass buffer metadata around, not keep things "alive". Why isn't enough with keeping the exporter alive?

Calling the deleter is meant to inform the exporter that the object is no longer needed by the consumer, and technically it is not the case here. In theory an early `deleter` call would not guarantee the exporter can survive until after the MPI call. In practice, though, it should be fine if `view` is going to hold a reference of the exporter to keep it alive.

Reply:
The `view` struct does hold a reference to the exporter. That should be enough to keep the exporter alive. And then we can get rid of the capsule and the DLPack stuff. Thus your changes should be reverted/discarded.

Reply:
Yup reverted in 4acb6bd.